src/engine/TransitivePath.cpp

```diff
@@ -658,119 +449,158 @@ void TransitivePath::computeTransitivePathRightBound(
-                                                       {inputCol});
-  // Create a copy of this.
-  //
-  // NOTE: The RHS used to be `std::make_shared<TransitivePath>()`, which is
-  // wrong because it first calls the copy constructor of the base class
-  // `Operation`, which  would then ignore the changes in `variableColumnMap_`
-  // made below (see `Operation::getInternallyVisibleVariableColumns` and
-  // `Operation::getExternallyVariableColumns`).
-  std::shared_ptr<TransitivePath> p = std::make_shared<TransitivePath>(
-      getExecutionContext(), _subtree, _leftIsVar, _rightIsVar, _leftSubCol,
-      _rightSubCol, _leftValue, _rightValue, Variable{_leftColName},
-      Variable{_rightColName}, _minDist, _maxDist);
-  if (isLeft) {
-    p->_leftSideTree = leftOrRightOp;
-    p->_leftSideCol = inputCol;
+template <size_t SUB_WIDTH>
+std::pair<TransitivePath::Map, std::vector<Id>>
+TransitivePath::setupMapAndNodes(const IdTable& sub,
+                                 const TransitivePathSide& startSide,
+                                 const TransitivePathSide& targetSide) const {
+  std::vector<Id> nodes;
+  Map edges = setupEdgesMap<SUB_WIDTH>(sub, startSide, targetSide);
+
+  // id -> var|id
+  if (startSide.isBound()) {
+    nodes.push_back(std::get<Id>(startSide.value));
+    // var -> var
   } else {
-    p->_rightSideTree = leftOrRightOp;
-    p->_rightSideCol = inputCol;
+    nodes = setupNodesVector<SUB_WIDTH>(sub, startSide.subCol);
+    if (_minDist == 0) {
+      std::vector<Id> targetNodes =
+          setupNodesVector<SUB_WIDTH>(sub, targetSide.subCol);
+      nodes.insert(nodes.end(), targetNodes.begin(), targetNodes.end());
+    }
   }
 
-  // Note: The `variable` in the following structured binding is `const`, even
-  // if we bind by value. We deliberately make one unnecessary copy of the
-  // `variable` to keep the code simpler.
-  for (auto [variable, columnIndexWithType] :
-       leftOrRightOp->getVariableColumns()) {
-    ColumnIndex columnIndex = columnIndexWithType.columnIndex_;
-    if (columnIndex != inputCol) {
-      if (columnIndex > inputCol) {
-        columnIndexWithType.columnIndex_++;
-        p->_variableColumns[variable] = columnIndexWithType;
-      } else {
-        columnIndexWithType.columnIndex_ += 2;
-        p->_variableColumns[variable] = columnIndexWithType;
-      }
-      p->_resultWidth++;
+  return std::make_pair(edges, nodes);
+}
+
+// _____________________________________________________________________________
+template <size_t SUB_WIDTH>
+TransitivePath::Map TransitivePath::setupEdgesMap(
+    const IdTable& dynSub, const TransitivePathSide& startSide,
+    const TransitivePathSide& targetSide) {
+  const IdTableView<SUB_WIDTH> sub = dynSub.asStaticView<SUB_WIDTH>();
+  Map edges;
+
+  for (size_t i = 0; i < sub.size(); i++) {
+    // checkTimeoutHashSet();
+    Id startId = sub(i, startSide.subCol);
+    Id targetId = sub(i, targetSide.subCol);
```

Comment:
You can upfront only get a view on the columns of the input
(`decltype(auto) startCol = sub.getColumn(startSide.subCol)`)
And then you can simply iterate over those columns (more efficient, and maybe also simpler code).